ngtsc as a plugin!

alxhub opened the pull request:
It's happening!!

This PR implements `NgTscPlugin` on top of some refactored compiler APIs. Hopefully this will allow integration with `ts_library` in the near future.

There's lots of rough edges left here, but it looks promising already!

alxhub committed:
feat(compiler-cli): implement NgTscPlugin on top of the NgCompiler API

This commit implements an experimental integration with tsc_wrapped, where
it can load the Angular compiler as a plugin and perform Angular
transpilation at a user's request.

This is an alternative to the current ngc_wrapped mechanism, which is a fork
of tsc_wrapped from several years ago. tsc_wrapped has improved
significantly since then, and this feature will allow Angular to benefit
from those improvements.

Currently the plugin API between tsc_wrapped and the Angular compiler is a
work in progress, so NgTscPlugin does not yet implement any interfaces from
@bazel/typescript (the home of tsc_wrapped). Instead, an interface is
defined locally to guide this standardization.

alexeagle commented on packages/compiler-cli/src/ngtsc/tsc_plugin.ts:
```diff
@@ -6,77 +6,88 @@
+import {PluginCompilerHost} from '@bazel/typescript/internal/tsc_wrapped/plugin_api';
 import * as ts from 'typescript';
 
-import {SyntheticFilesCompilerHost} from './synthetic_files_compiler_host';
```
I think you can delete this file now

alexeagle commented on packages/compiler-cli/src/ngtsc/tsc_plugin.ts:
```diff
@@ -6,77 +6,88 @@
+  getNextProgram(): ts.Program;
+
+  prepareEmit(): {
+    transformers: {
```
could you use the type `ts.CustomTransformers` here?

alxhub replied:
Done.

alexeagle commented on packages/compiler-cli/src/ngtsc/tsc_plugin.ts:
```diff
@@ -6,77 +6,88 @@
+ * enable progress to be made in parallel, the upstream interface isn't implemented directly.
+ * Instead, `TscPlugin` here is structurally assignable to what tsc_wrapped expects.
+ */
+interface TscPlugin {
```
could you add a member `name = 'ngtsc'` so that it matches the interface LitPlugin introduced for gathering diagnostics?

alxhub replied:
Done.

alexeagle commented on packages/compiler-cli/src/ngtsc/tsc_plugin.ts:
```diff
@@ -6,77 +6,88 @@
+ */
+interface TscPlugin {
+  wrapHost(
+      host: PluginCompilerHost, inputFiles: ReadonlyArray<string>,
```
why does this need to be a `PluginCompilerHost`? in our work last week we wrapped a `ts.CompilerHost` which was the only thing constructed at this point

alxhub replied:
I've refactored it to take a `ts.CompilerHost` plus the optional `fileNameToModuleName`.

alexeagle commented on packages/compiler-cli/src/ngtsc/tsc_plugin.ts:
```diff
@@ -6,77 +6,88 @@
+interface TscPlugin {
+  wrapHost(
+      host: PluginCompilerHost, inputFiles: ReadonlyArray<string>,
+      options: ts.CompilerOptions): ts.CompilerHost;
```
and I think it needs to return a PluginCompilerHost so we can locate the inputFiles when we construct a ts.Program

alxhub replied:
Done.

alexeagle commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -0,0 +1,809 @@
+    if (file === undefined) {
+      return this.diagnostics;
+    } else {
+      return this.diagnostics.filter(diag => diag.file === file);
```
this is broken - diag.file is the .ng.html file but file is the .ts file

alxhub replied:
Fixed.

JoostK commented on packages/compiler-cli/src/ngtsc/core/src/host.ts:
```diff
@@ -0,0 +1,248 @@
+                                               undefined;
+  }
+
+  // Excluded are 'getSourceFile' and 'fileExists', which are actually implemented by NgCompilerHost
```
Did you consider collecting all delegate methods into an object literal, and copying over the ones that are present in the delegate host but not on the prototype as custom methods? This way, we can avoid these exclusions _and_ fully omit setting properties that the delegate doesn't have anyways.

Something like this:

```ts
type RequiredCompilerHostDelegations = {
   [M in keyof Required<ExtendedTsCompilerHost>]: null;
}

const delegateMethods: RequiredCompilerHostDelegations  = {
  createHash: null,
  directoryExists: null,
  // ... etc.
}

export class DelegatingCompilerHost implements ExtendedTsCompilerHost {
  constructor(protected delegate: ExtendedTsCompilerHost) {
    for (const delegateMethod of Object.keys(delegateMethods)) {
      if (!(delegateMethod in this) && (delegate as any)[delegateMethod] !== undefined) {
       (this as any)[delegateMethod] = (delegate as any)[delegateMethod].bind(delegate);
      }
    }
  }

}
```

alxhub replied:
Yes, but I prefer the explicitness of adding each method directly. The above won't actually type-check because dynamically adding the methods won't satisfy the interface implementation, so you'll need to add a lie to the type system to address that.

JoostK reviewed (commented):
Stellar work! I like how the compiler objects are nicely separated and how the shim generator host has been centralized into the `NgCompilerHost`. Some minor comments for your consideration.

alxhub commented on packages/compiler-cli/src/ngtsc/tsc_plugin.ts:
```diff
@@ -6,77 +6,88 @@
+
+  get compiler(): NgCompiler {
+    if (this._compiler === null) {
+      throw new Error('Lifecycle error: getDiagnostics() called before setupCompilation()');
```
Fixed :)

alxhub commented on packages/compiler-cli/src/ngtsc/core/src/host.ts:
```diff
@@ -0,0 +1,248 @@
+ * optional).
+ */
+export type RequiredCompilerHostDelegations = {
+  [M in keyof Required<ExtendedTsCompilerHost>]: ExtendedTsCompilerHost[M]
```
👍

alxhub commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -0,0 +1,809 @@
+          injectableRegistry, this.closureCompilerEnabled),
+      // TODO(alxhub): understand why the cast here is necessary (something to do with `null`
+      // not
+      // being assignable to `unknown` when wrapped in `Readonly`).
```
👍

alxhub commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -0,0 +1,809 @@
+          compilation.traitCompiler, compilation.reflector, importRewriter,
+          compilation.defaultImportTracker, compilation.isCore, this.closureCompilerEnabled),
+      // clang-format off
+      aliasTransformFactory(compilation.traitCompiler.exportStatements) as ts.TransformerFactory<ts.SourceFile>,
```
Done. Much cleaner!

JoostK commented on packages/compiler-cli/src/ngtsc/core/api.ts:
```diff
@@ -0,0 +1,360 @@
+/**
+ * A host backed by a build system which has a unified view of the module namespace.
+ *
+ * Such a build system is able to map be supports the `moduleNameToFileName` method provided by
```
I don't understand this comment, `able to map be supports the` seems incorrect.

JoostK commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -0,0 +1,810 @@
+      return this.diagnostics.filter(diag => {
+        if (diag.file === file) {
+          return true;
+        } else if (isTemplateDiagnostic(diag) && diag.componentFile === file) {
```
This would be worth having in a unit test, I don't think we have such a test at the moment.

alxhub replied:
I agree, but I'm gonna defer it because we don't have any testing infrastructure that invokes the compiler at this level yet.

JoostK reviewed (approved):
LGTM, with some more minor comments

alxhub commented:
This is blocked on the upstream g3 work to standardize the plugin API, as well as #34887 landing first.

alexeagle commented:
Should be unblocked for g3sync now - I commented out the relevant bit in tsc_wrapped

alxhub committed:
refactor(ivy): compute ignoreFiles for compilation on initialization

This commit moves the calculation of `ignoreFiles` - the set of files to be
ignored by a consumer of the `NgCompiler` API - from its `prepareEmit`
operation to its initialization. It's now available as a field on
`NgCompiler`.

This will allow a consumer to skip gathering diagnostics for `ignoreFiles`
as well as skip emit.

alxhub commented:
[Presubmit](http://test/OCL:290825946:BASE:293477683:1580948480305:c5a0aede)

alxhub commented:
Caretaker: I cannot force g3 status ("This SHA and context has reached the maximum number of statuses" error). Presubmits are good (only Ivy failures are screenshot diffs & timeouts).

LayZeeDK commented:
So this is something Bazel-related? Would anyone care to explain what the big deal is? 🤔
